ddtrace/internal/_encoding.pyx

```diff
@@ -46,120 +47,327 @@ cdef inline int PyBytesLike_Check(object o):
     return PyBytes_Check(o) or PyByteArray_Check(o)
 
 
-cdef inline int pack_bytes(msgpack_packer *pk, char *bytes, Py_ssize_t l):
+cdef inline int pack_bytes(msgpack_packer *pk, char *bs, Py_ssize_t l):
     cdef int ret
-    cdef dict d
+
     ret = msgpack_pack_raw(pk, l)
     if ret == 0:
-        ret = msgpack_pack_raw_body(pk, bytes, l)
+        ret = msgpack_pack_raw_body(pk, bs, l)
     return ret
 
 
-cdef class Packer(object):
-    """
-    MessagePack Packer
+cdef inline int pack_number(msgpack_packer * pk, object n):
+    if n is None:
+        return msgpack_pack_nil(pk)
+
+    if PyLong_Check(n):
+        # PyInt_Check(long) is True for Python 3.
+        # So we should test long before int.
+        try:
+            if n > 0:
+                return msgpack_pack_unsigned_long_long(pk, <unsigned long long> n)
+            return msgpack_pack_long_long(pk, <long long> n)
+        except OverflowError as oe:
+            raise OverflowError("Integer value out of range")
+
+    if PyInt_Check(n):
+        return msgpack_pack_long(pk, <long> n)
+
+    if PyFloat_Check(n):
+        return msgpack_pack_double(pk, <double> n)
+
+    raise TypeError("Unhandled numeric type: %r" % type(n))
+
+
+cdef inline int pack_text(msgpack_packer *pk, object text):
+    cdef Py_ssize_t L
+    cdef int ret
+
+    if text is None:
+        return msgpack_pack_nil(pk)
+
+    if PyBytesLike_Check(text):
+        L = len(text)
+        if L > ITEM_LIMIT:
+            PyErr_Format(ValueError, b"%.200s object is too large", Py_TYPE(text).tp_name)
+        ret = msgpack_pack_raw(pk, L)
+        if ret == 0:
+            ret = msgpack_pack_raw_body(pk, <char *> text, L)
+        return ret
+
+    if PyUnicode_Check(text):
+        IF PY_MAJOR_VERSION >= 3:
+            ret = msgpack_pack_unicode(pk, text, ITEM_LIMIT)
+            if ret == -2:
+                raise ValueError("unicode string is too large")
+        ELSE:
+            text = PyUnicode_AsEncodedString(text, "utf-8", NULL)
+            L = len(text)
+            if L > ITEM_LIMIT:
+                raise ValueError("unicode string is too large")
+            ret = msgpack_pack_raw(pk, L)
+            if ret == 0:
+                ret = msgpack_pack_raw_body(pk, <char *> text, L)
+        return ret
+
+    raise TypeError("Unhandled text type: %r" % type(text))
+
 
-    usage::
+cdef inline int array_prefix_size(int l):
+    if l < 16:
+        return 1
+    elif l < (2<<16):
+        return 3
+    return 5
 
-        packer = Packer()
-        astream.write(packer.pack_trace(trace))
-        astream.write(packer.pack_traces(traces))
 
-    Packer's constructor has some keyword arguments:
+cdef class StringTable(object):
+    cdef dict _index
 
-    :param callable default:
-        Convert user type to builtin type that Packer supports.
-        See also simplejson's document.
-    """
+    def __init__(self):
+        self._index = {"": 0}
```

Comment:
Why do we need to add `"": 0` to the index during startup?

Reply:
> does it make sense to pre-fill this table with other common values we expect (almost) every trace to have?

Yes it does! We were discussing this with @Yun-Kim and agreed this could be a further optimisation after this initial implementation. In the meantime we can try to figure out which strings should be pre-filled, or if it would make sense to make it an "MFU string table".

Reply:
MFU would make a lot of sense, since a given application is probably going to have high span resource/name, and tag name reuse at least.

Reply:
Should we make this enhancement in a different PR? v0.5 is still not enabled at this point.